[WebGPU] GPUDevice.lost promise is never resolved

mwyrzykowski opened the pull request:
#### 65f2377f0c0973c8984aa4eda1a4cee8a173dcda
<pre>
[WebGPU] GPUDevice.lost promise is never resolved
<a href="https://bugs.webkit.org/show_bug.cgi?id=253086">https://bugs.webkit.org/show_bug.cgi?id=253086</a>
&lt;radar://106037229&gt;

Reviewed by Myles C. Maxfield.

Adopt new agreed upon API for device lost.

CTS test is not enabled because we fail requestDevice now due to
maxFragmentCombinedOutputResources being added to GPULimits.

Locally the requestDevice:invalid test is passing with this change.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::lost):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
(WebCore::GPUDevice::lost): Deleted.
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.cpp:
(WebCore::GPUDeviceLostInfo::reason const):
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.h:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostReason.h:
(WebCore::convertToBacking):
* Source/WebCore/Modules/WebGPU/GPUDeviceLostReason.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::resolveDeviceLostPromise):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDeviceLostInfo.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDeviceLostReason.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::~Device):
(WebGPU::Device::loseTheDevice):
(WebGPU::Device::setDeviceLostCallback):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::resolveDeviceLostPromise):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::resolveDeviceLostPromise):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:

Canonical link: <a href="https://commits.webkit.org/263100@main">https://commits.webkit.org/263100@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/67fa019f5975b364d843a2be35be382d9d4032a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3626 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3767 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5005 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3854 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3668 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3112 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3623 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3870 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3227 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4827 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1376 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3200 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3268 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3177 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4597 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3637 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2942 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3177 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3200 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/878 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3205 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3456 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/Modules/WebGPU/GPUDevice.h:
```diff
@@ -41,6 +41,12 @@
 #include <wtf/text/WTFString.h>
 
+namespace PAL {
+namespace WebGPU {
```
`namespace PAL::WebGPU {`

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUAdapter.h:
```diff
@@ -48,6 +48,8 @@ class Adapter : public RefCounted<Adapter> {
     bool isFallbackAdapter() const { return m_isFallbackAdapter; }
 
     virtual void requestDevice(const DeviceDescriptor&, CompletionHandler<void(RefPtr<Device>&&)>&&) = 0;
+    void setInvalid() { m_valid = false; }
```
maybe `void makeInvalid()`?

litherum commented on Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDeviceLostReason.h:
```diff
@@ -32,6 +32,7 @@ namespace PAL::WebGPU {
 
 enum class DeviceLostReason : uint8_t {
     Destroyed,
+    Unknown
```
Do we want a trailing comma? I usually do them.

litherum commented on Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
```diff
@@ -59,27 +59,35 @@
-GPUDevice::~GPUDevice() = default;
+GPUDevice::~GPUDevice()
+{
+    destroy();
```
Are we sure we want the destructor to resolve a promise? I don't think we do.

mwyrzykowski replied:
With the introduction of `destruct` which is called from the destructor, I will only have destroy resolve the promise.

litherum commented on Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
```diff
@@ -108,7 +116,22 @@ Ref<GPUQueue> GPUDevice::queue() const
+    m_destroyed = true;
     m_backing->destroy();
+    auto ref = GPUDeviceLostInfo::create(PAL::WebGPU::DeviceLostInfo::create(PAL::WebGPU::DeviceLostReason::Unknown, ""_s));
+    m_lostPromise->resolve(WTFMove(ref));
```
Is this right? 🤔

mwyrzykowski replied:
I think so, it passes the test. Is there an alternative?

litherum commented on Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
```diff
@@ -59,27 +59,35 @@
+    , m_adapter(adapter)
 {
+    if (!adapter->isValid())
+        makeInvalid();
```
Why isn't this validity logic done inside WebGPU.framework? These GPUFoo classes are supposed to be thin wrappers.

mwyrzykowski replied:
I will try to move it to WebGPU.framework, good idea!

mwyrzykowski replied:
I think the promise can be asynchronously resolved in `GPUDevice::lost(...)` to avoid having this logic here, I will give that a try.

mwyrzykowski replied:
@litherum I moved the changes down into WebGPU.framework, leveraging what was already there. It seems to work, but needs more testing to make sure we don't retain the GPUDevice due to retaining it in `resolveDeviceLostPromise`.

Please let me know if you have any further comments, otherwise I will investigate the possible retain cycle prior to merging

mwyrzykowski replied:
Oh I think I can just use `WeakPtr` instead of `Ref` in the callback, please let me know what you think.

mwyrzykowski commented on Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
```diff
@@ -111,6 +112,23 @@ void GPUDevice::destroy()
+        return m_lostPromise;
+
+    m_waitingForDeviceLostPromise = true;
+    m_backing->resolveDeviceLostPromise([weakThis = WeakPtr { *this }] (PAL::WebGPU::DeviceLostReason reason) {
```
I used `WeakPtr` here to avoid a retain cycle where GPUDevice's destructor would never execute

mwyrzykowski committed:
[WebGPU] GPUDevice.lost promise is never resolved
https://bugs.webkit.org/show_bug.cgi?id=253086
<radar://106037229>

Reviewed by Myles C. Maxfield.

Adopt new agreed upon API for device lost.

CTS test is not enabled because we fail requestDevice now due to
maxFragmentCombinedOutputResources being added to GPULimits.

Locally the requestDevice:invalid test is passing with this change.

* Source/WebCore/Modules/WebGPU/GPUDevice.cpp:
(WebCore::GPUDevice::lost):
* Source/WebCore/Modules/WebGPU/GPUDevice.h:
(WebCore::GPUDevice::lost): Deleted.
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.cpp:
(WebCore::GPUDeviceLostInfo::reason const):
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.h:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostInfo.idl:
* Source/WebCore/Modules/WebGPU/GPUDeviceLostReason.h:
(WebCore::convertToBacking):
* Source/WebCore/Modules/WebGPU/GPUDeviceLostReason.idl:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.cpp:
(PAL::WebGPU::DeviceImpl::resolveDeviceLostPromise):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUDeviceImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDevice.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDeviceLostInfo.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUDeviceLostReason.h:
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::~Device):
(WebGPU::Device::loseTheDevice):
(WebGPU::Device::setDeviceLostCallback):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.cpp:
(WebKit::RemoteDevice::resolveDeviceLostPromise):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.h:
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteDevice.messages.in:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.cpp:
(WebKit::WebGPU::RemoteDeviceProxy::resolveDeviceLostPromise):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteDeviceProxy.h:

Canonical link: https://commits.webkit.org/263100@main
